### ISSUE TITLE
no-underscore-dangle option to only report declarations

### ISSUE THREAD

**Issue by privatenumber** [eslint/eslint#13945]
**What rule do you want to change?**
[no-underscore-dangle](https://eslint.org/docs/rules/no-underscore-dangle)

**Does this change cause the rule to produce more or fewer warnings?**
Depends on configuration. The change would add more nuance/control over what it errors/warns on.

**How will the change be implemented? (New option, new default behavior, etc.)?**
New option. Maybe `allowDeclaration` defaulting to `false`.

Or `allowAccess` that would replace `allowAfterThis`, `allowAfterSuper`, and `allowAfterThisConstructor`:
```
{
	allowAccess: true,

	allowAccess: {
		afterThis: false,
		afterSuper: false,
		afterThisConstructor: false
	}
}
```

Open to other ideas.

**Please provide some example code that this change will affect:**

```js
/*eslint no-underscore-dangle: ["error", { "allowDeclaration": false }]*/

var foo_; // warn
const _bar; // warn
function foo (_bar) {} // warn
const fooBar = {
	_foo: 1, // warn
	bar_: 1, // warn
};
class Foo {
  _bar() {}
}


foo._bar(); // OK
var a = this.foo_; // OK
this._bar(); // OK
```

**What does the rule currently do for this code?**
It reports error on underscore usage from imported modules I don't have control over.

**What will the rule do after it's changed?**
It will only disallow underscore usage in my own code, but will allow accessing dangling underscore properties from 3rd-party modules.

**Are you willing to submit a pull request to implement this change?**
Yes

**Comment by mdjermanovic** [comment]
Hi @privatenumber, thanks for the issue!

Per our [rules policies](https://eslint.org/blog/2020/05/changes-to-rules-policies), stylistic rules are frozen, so we won't be adding new options to the `no-underscore-dangle` rule.

You can always make a custom rule based on the existing one. Tools like [eslint-rule-extender](https://github.com/kaicataldo/eslint-rule-extender) can be helpful if the custom rule modifies the behavior of another rule, e.g. if it filters out some errors:

```js
const ruleExtender = require('eslint-rule-extender');
const eslint = require('eslint');

const noUnderscoreDangle = new eslint.Linter().getRules().get('no-underscore-dangle');

// new rule
module.exports = ruleExtender(noUnderscoreDangle, {
  reportOverrides(meta) {
    return meta.node.type !== "MemberExpression"
  }
});
```
